### PR TITLE
chore: Add tagging for migration to `eks:eks-cluster-name`

### PR DESCRIPTION
**Pull request by jonathan-innis**
Fixes #N/A 

**Description**

This change ensures that resources that are managed by Karpenter contain the new `eks:eks-cluster-name` tag so that users can write identity policies for Karpenter that are based on [EKS Pod Identity Template Parameters](https://docs.aws.amazon.com/eks/latest/userguide/pod-identities.html).

**How was this change tested?**

`make presubmit`
`/karpenter snapshot`

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by jonathan-innis — COMMENTED** [review]
/karpenter snapshot

**Review by jonathan-innis — COMMENTED** [review]
/karpenter snapshot

**Review by jonathan-innis — COMMENTED** [review]
/karpenter snapshot

File: pkg/controllers/nodeclaim/tagging/controller.go [hunk 1]
```diff
@@ -95,8 +99,9 @@ func (c *Controller) Register(_ context.Context, m manager.Manager) error {
 
 func (c *Controller) tagInstance(ctx context.Context, nc *karpv1.NodeClaim, id string) error {
 	tags := map[string]string{
-		v1.TagName:      nc.Status.NodeName,
-		v1.TagNodeClaim: nc.Name,
+		v1.TagName:              nc.Status.NodeName,
+		v1.TagNodeClaim:         nc.Name,
+		v1.EKSClusterNameTagKey: options.FromContext(ctx).ClusterName,
```
Comment (njtran) [inline pkg/controllers/nodeclaim/tagging/controller.go:104]: Do we pass the cluster name into this context? I know we do normally, but want to make sure it's at the right stage where the settings are already parsed in

File: pkg/controllers/nodeclaim/tagging/controller.go [hunk 2]
```diff
@@ -120,7 +125,9 @@ func (c *Controller) tagInstance(ctx context.Context, nc *karpv1.NodeClaim, id s
 
 func isTaggable(nc *karpv1.NodeClaim) bool {
 	// Instance has already been tagged
-	if val := nc.Annotations[v1.AnnotationInstanceTagged]; val == "true" {
+	instanceTagged := nc.Annotations[v1.AnnotationInstanceTagged]
+	clusterNameTagged := nc.Annotations[v1.AnnotationClusterNameTaggedCompatability]
+	if instanceTagged == "true" && clusterNameTagged == "true" {
```
Comment (njtran) [inline pkg/controllers/nodeclaim/tagging/controller.go:130]: If only one of these is tagged, do we idempotently add the one that's already in or does that fail?

File: pkg/controllers/nodeclaim/tagging/suite_test.go [hunk 3]
```diff
@@ -220,9 +221,12 @@ var _ = Describe("TaggingController", func() {
 				Expect(instanceTags).To(HaveKeyWithValue(tag, value))
 			}
 		},
-		Entry("with only karpenter.k8s.aws/nodeclaim tag", v1.TagName),
-		Entry("with only Name tag", v1.TagNodeClaim),
-		Entry("with both Name and karpenter.k8s.aws/nodeclaim tags"),
-		Entry("with nothing to tag", v1.TagName, v1.TagNodeClaim),
+		Entry("with the karpenter.sh/nodeclaim tag", v1.TagName, v1.EKSClusterNameTagKey),
```
Comment (njtran) [inline pkg/controllers/nodeclaim/tagging/suite_test.go:224]: is this an expected change? from karpenter.k8s.aws to karpenter.sh?

File: pkg/fake/iamapi.go [hunk 4]
```diff
@@ -112,6 +113,31 @@ func (s *IAMAPI) DeleteInstanceProfileWithContext(_ context.Context, input *iam.
...
+						foundTag = true
+						break
+					}
+				}
+				if !foundTag {
+					profile.Tags = append(profile.Tags, tag)
+				}
+			}
+			return nil, nil
+		}
```
Comment (njtran) [inline pkg/fake/iamapi.go:136]: Is there a way we can nest this less?

File: test/suites/integration/tags_test.go [hunk 5]
```diff
@@ -55,14 +56,14 @@ var _ = Describe("Tags", func() {
 				return ni.NetworkInterfaceId
 			})...)
 
-			Expect(tagMap(instance.Tags)).To(HaveKeyWithValue("TestTag", "TestVal"))
+			Expect(instance.Tags).To(ContainElement(&ec2.Tag{Key: lo.ToPtr("TestTag"), Value: lo.ToPtr("TestVal")}))
 			for _, volume := range volumes {
-				Expect(tagMap(volume.Tags)).To(HaveKeyWithValue("TestTag", "TestVal"))
+				Expect(volume.Tags).To(ContainElement(&ec2.Tag{Key: lo.ToPtr("TestTag"), Value: lo.ToPtr("TestVal")}))
```
Comment (njtran) [inline test/suites/integration/tags_test.go:61]: Does asserting that we have the element's pointer get us into any issues here, or does containElement go into the address for us

File: test/suites/integration/tags_test.go [hunk 6]
```diff
@@ -82,7 +83,42 @@ var _ = Describe("Tags", func() {
...
+				&iam.Tag{Key: lo.ToPtr(fmt.Sprintf("kubernetes.io/cluster/%s", env.ClusterName)), Value: lo.ToPtr("owned")},
+				&iam.Tag{Key: lo.ToPtr(v1.LabelNodeClass), Value: lo.ToPtr(nodeClass.Name)},
+				&iam.Tag{Key: lo.ToPtr(v1.EKSClusterNameTagKey), Value: lo.ToPtr(env.ClusterName)},
+			))
+		})
+		It("should tag managed instance profiles with the eks:eks-cluster-name tag key after restart", func() {
+			env.ExpectCreated(nodeClass)
+			env.EventuallyExpectInstanceProfileExists(env.GetInstanceProfileName(nodeClass))
+
+			_, err := env.IAMAPI.UntagInstanceProfile(&iam.UntagInstanceProfileInput{
```
Comment (njtran) [inline test/suites/integration/tags_test.go:103]: Is this another permission in our cloud formation that we need to give explicitly to our e2es?

File: test/suites/integration/tags_test.go [hunk 7]
```diff
@@ -82,7 +83,42 @@ var _ = Describe("Tags", func() {
...
+			Expect(err).ToNot(HaveOccurred())
+
+			// Restart Karpenter to flush the instance profile cache and to trigger re-tagging of the instance profile
+			env.EventuallyExpectKarpenterRestarted()
+
+			Eventually(func(g Gomega) {
+				out, err := env.IAMAPI.GetInstanceProfile(&iam.GetInstanceProfileInput{
+					InstanceProfileName: lo.ToPtr(env.GetInstanceProfileName(nodeClass)),
+				})
+				fmt.Println(out, err)
```
Comment (njtran) [inline test/suites/integration/tags_test.go:118]: Remove?

File: test/suites/integration/tags_test.go [hunk 8]
```diff
@@ -93,19 +129,17 @@ var _ = Describe("Tags", func() {
...
-			nodeName := client.ObjectKeyFromObject(node)
+			nodeClaim := env.ExpectNodeClaimCount("==", 1)[0]
 
 			Eventually(func(g Gomega) {
-				node = &corev1.Node{}
-				g.Expect(env.Client.Get(env.Context, nodeName, node)).To(Succeed())
-				g.Expect(node.Annotations).To(HaveKeyWithValue(v1.AnnotationInstanceTagged, "true"))
-			}, time.Minute)
+				g.Expect(env.Client.Get(env.Context, client.ObjectKeyFromObject(nodeClaim), nodeClaim)).To(Succeed())
+				g.Expect(nodeClaim.Annotations).To(HaveKeyWithValue(v1.AnnotationInstanceTagged, "true"))
```
Comment (njtran) [inline test/suites/integration/tags_test.go:136]: do we also need to check if cluster tagged?

File: test/suites/integration/tags_test.go [hunk 9]
```diff
@@ -132,23 +166,51 @@ var _ = Describe("Tags", func() {
...
+			}, time.Minute).Should(Succeed())
+
+			_, err := env.EC2API.DeleteTags(&ec2.DeleteTagsInput{
+				Resources: []*string{lo.ToPtr(env.ExpectParsedProviderID(node.Spec.ProviderID))},
+				Tags:      []*ec2.Tag{{Key: lo.ToPtr(v1.EKSClusterNameTagKey)}},
+			})
+			Expect(err).ToNot(HaveOccurred())
+
+			By(fmt.Sprintf("Removing the %s annotation to re-trigger tagging", v1.AnnotationClusterNameTaggedCompatability))
```
Comment (njtran) [inline test/suites/integration/tags_test.go:199]: Shouldn't we first check and expect that it exists so we know we're removing it?

File: pkg/controllers/nodeclaim/tagging/controller.go (hunk 1 above)
Comment (jonathan-innis) [reply pkg/controllers/nodeclaim/tagging/controller.go:104]: > stage where the settings are already parsed in

Yep, they're here

File: pkg/controllers/nodeclaim/tagging/controller.go (hunk 2 above)
Comment (jonathan-innis) [reply pkg/controllers/nodeclaim/tagging/controller.go:130]: > do we idempotently add the one that's already in or does that fail

It doesn't fail, we would just only add the tags that aren't there on the resource and not update the others

File: pkg/controllers/nodeclaim/tagging/suite_test.go (hunk 3 above)
Comment (jonathan-innis) [reply pkg/controllers/nodeclaim/tagging/suite_test.go:224]: Yes, this was wrong originally -- just updating to make it correct

File: test/suites/integration/tags_test.go (hunk 5 above)
Comment (jonathan-innis) [reply test/suites/integration/tags_test.go:61]: Nope, it checks the address so we don't run into any issues here. The tests pass :)

File: test/suites/integration/tags_test.go (hunk 6 above)
Comment (jonathan-innis) [reply test/suites/integration/tags_test.go:103]: Yep, already added :)

File: test/suites/integration/tags_test.go (hunk 7 above)
Comment (jonathan-innis) [reply test/suites/integration/tags_test.go:118]: Good call. Oopsie

File: test/suites/integration/tags_test.go (hunk 9 above)
Comment (jonathan-innis) [reply test/suites/integration/tags_test.go:198]: Sure, can do

File: pkg/fake/iamapi.go (hunk 4 above)
Comment (jonathan-innis) [reply pkg/fake/iamapi.go:136]: Teach me your ways `lo` god 😄 

```go
profile.Tags = lo.UniqBy(append(input.Tags, profile.Tags...), func(t *iam.Tag) string {
  return lo.FromPtr(t.Key)
})
```

**Review by jonathan-innis — COMMENTED** [review]
/karpenter snapshot

**Review by rschalo — APPROVED** [review]

**Pull request merged**